src/main/scala/shine/OpenCL/compilation/HostManagedBuffers.scala

```diff
@@ -0,0 +1,271 @@
+package shine.OpenCL.compilation
+
+import shine.DPIA._
+import shine.DPIA.Phrases._
+import shine.DPIA.Semantics.OperationalSemantics
+import shine.DPIA.Types._
+import shine.DPIA.primitives.{imperative => dpia}
+import shine.OpenCL._
+import shine.OpenCL.primitives.{imperative => ocl}
+
+import scala.annotation.tailrec
+import scala.collection.{immutable, mutable}
+
+object HostManagedBuffers {
+  def populate(params: immutable.Seq[Identifier[ExpType]], outParam: Identifier[AccType])
+  : Phrase[CommType] => Phrase[CommType] = { p =>
+    def outsideParams() = mutable.Set[Identifier[_ <: PhraseType]]() ++ (params :+ outParam)
+    val worstCasePrevious =
+      Metadata(outsideParams(), outsideParams(), outsideParams(), outsideParams())
+    val managed = mutable.Map[Identifier[_ <: PhraseType], AccessFlags]()
+    insertHostExecutions(worstCasePrevious, outsideParams().toSet, managed, p)._1 |>
+    insertManagedBuffers(managed)
+  }
+
+  final case class HostExecution(env: Map[Identifier[_ <: PhraseType], AccessFlags],
```

Comment:
What is a `HostExecution`? Why is this part of the `HostManagedBuffers` object?

Reply:
A `HostExecution` is doing some processing on the host and basically can be delegated to the plain C code generator without being aware of the host/device setup. I do not have a very good name in mind but I see it as launching a C kernel on the host (in constrast to launching an OpenCL kernel on the device).

Reply:
and the primitive can be extracted from the `HostManagedBuffers` object if/when the name is satisfying, this happened while coding as I first planned to eliminate this primitive before codegen.

Reply:
Thanks for pulling this out and adding an explaining comment.